components/utils.py

```diff
@@ -310,26 +336,72 @@ def yum_install(source):
         source_name = subprocess.check_output(
             ['rpm', '-qp', source_path]).strip()
 
-    ctx.logger.info('Checking whether {0} is already installed...'.format(
-        source_path.rstrip('\n\r')))
-    installed = run(['rpm', '-q', source_name], ignore_failures=True)
-    if installed.returncode == 0:
-        ctx.logger.info('Package {0} is already installed.'.format(source))
-        return
+        ctx.logger.info('Checking whether {0} is already installed...'.format(
+            source_path))
+        installed_rpm = _get_installed_rpm_source(source_name)
+        if installed_rpm == source_name:
+            ctx.logger.info('Package {0} is already installed.'.format(source))
+            return
+
+        if is_upgrade() or is_rollback():
+            _remove_existing_rpm_package(source_name)
 
     ctx.logger.info('yum installing {0}...'.format(source_path))
     sudo(['yum', 'install', '-y', source_path])
 
 
+def _remove_existing_rpm_package(source_name):
```

Comment:
considering the set of functions we use to handle rpm files, i'd create a class called `RpmHandler` and instantiate it once from the yum_install function to keep the code clean.